apiserver/raft.go

```diff
@@ -79,10 +83,13 @@ func (m *raftMediator) ApplyLease(cmd []byte) error {
 			m.logger.Tracef("Applied Lease with command %s in time: %v", string(cmd), elapsed)
 		}
 
-	case <-m.clock.After(ApplyLeaseTimeout):
-		elapsed := m.clock.Now().Sub(start)
+	case <-ctx.Done():
+		m.logger.Errorf("Context was marked as done whilst waiting for enqueue: %v", m.clock.Now().Sub(start))
+		msg := fmt.Sprintf("Apply lease context deadline exceeded: %s", ctx.Err())
+		return apiservererrors.NewDeadlineExceededError(msg)
 
-		m.logger.Errorf("Applying Lease timed out, waiting for enqueue: %v", elapsed)
+	case <-m.clock.After(ApplyLeaseTimeout):
+		m.logger.Errorf("Applying Lease timed out, waiting for enqueue: %v", m.clock.Now().Sub(start))
```

Comment:
Do we care to add a ` if err, gotErr := <- done; gotErr {}` check here or do we just assume that the timeout error masks any apply errors (think of a scenario where apply actually completes at t+1ms past ApplyLeaseTimeout but we follow the clock.After select branch)?

Reply:
No, because the upper bound of the ApplyLeaseTimeout is large enough that if it did get through it would be almost out of date anyway. It can retry with a better expiry or someone else can steal it, which gives a better outcome.